### ISSUE TITLE
Access children `DataType` or return-type in `ScalarUDFImpl::invoke`

### ISSUE THREAD

**Issue by joseph-isaacs** [apache/datafusion#12819]
### Is your feature request related to a problem or challenge?

I am trying to create a scalar UDF, pack, which operates on struct arrays. It packs many array into a struct array each with a distinct name 

```pack(("a", arr1), ("b", arr2), ...) -> struct([("a", arr1.data_type), ("b", arr2.data_type), ...])```

This has a data type dependent on the input type and nullability. In the method `ScalarUDFImpl::invoke` I want to return an a struct array with each field having the data type and nullability of the input, however the invoke function only gives the data type of the array not the nullability of the record batch or intermediate children expressions. 

I have returned this type information from `return_type_from_exprs`, I just need to access this in the stateless scalar udf impl.

### Describe the solution you'd like

I would like add a new `ScalarUDFImpl::invoke_with_data_type` (or `invoke_with_return_type`) method which is given both the evaluated children array (as previously) and also either the previously returned type (from `return_type_from_exprs`) or the arguments already passed to `return_type_from_exprs` which could be re-evaluated by invoke. I am open to either, I guess the former seems more performant. 

### Additional context

I believe this would be a small non-breaking, change, that I am happy to contribute.

Any ideas?

**Comment by findepi** [comment]
You may want to implement `ScalarUDFImpl::simplify` which is given nullability info.
From there you'd return a  new  `ScalarUDFImpl` instance with the nullability information stored on a field. 
When the information is already stored, subsequent `simplify` call (if any), should return original expression.
Let me know if it works for your case.

At some point we could probably rename `simplify` to `specialize` cc @alamb

**Comment by alamb** [comment]
The suggestion for using `simplify` is a good one 👍

**Comment by joseph-isaacs** [comment]
This would be the [(non-breaking) change](https://github.com/spiraldb/datafusion/pull/1/files) I am after

**Comment by joseph-isaacs** [comment]
@alamb or @findepi. It would be great to get a response here. Thanks

**Comment by alamb** [comment]
I just read this again (quickly)

I wonder if you could (re) compute the return type from the input types using the existing `return_type` method:

https://docs.rs/datafusion/latest/datafusion/logical_expr/trait.ScalarUDFImpl.html#tymethod.return_type

**Comment by joseph-isaacs** [comment]
Thanks for your reply.

What do you think about the above proposed changed?

When, in the creation/execution path, do you mean re-compute the `return_type`? At invoke time?

I need this type-info when calling invoke (invoke is given a `[ColumnarValue]`) which is missing the necessarily type information to reconstruct the type and I cannot call `return_type`, taking `[DataType]` since `[DataType]` is not around at invoke time.

**Comment by alamb** [comment]
> When, in the creation/execution path, do you mean re-compute the return_type? At invoke time?

Yes


> invoke is given a [ColumnarValue]) which is missing the necessarily type information to reconstruct the type 

I think you can get it from: https://docs.rs/datafusion/latest/datafusion/logical_expr/enum.ColumnarValue.html#method.data_type

**Comment by joseph-isaacs** [comment]
I am sorry for my previous response it was not quite what I was asking for.

I think _your_ proposed solution is?
```rust
fn invoke(&self, args: &[ColumnarValue]) -> DFResult<ColumnarValue> {
  let arg_types = args.iter().map(|cv| cv.data_type()).collect_vec();
  let return_type = self.return_type(&arg_types)?;
  ...
}
```

I was hoping to use the result of `return_type_from_exprs(self, &[Expr], &dyn ExprSchema, &[DataType])` (in invoke) since I call `expr.data_type_and_nullable(schema)?` on the inputs to get the correct nullability.

**Comment by alamb** [comment]
> I think your proposed solution is?

Yes that is what I was thinking

> I was hoping to use the result of return_type_from_exprs(self, &[Expr], &dyn ExprSchema, &[DataType]) (in invoke) since I call expr.data_type_and_nullable(schema)? on the inputs to get the correct nullability.


Ah, I see -- the nullability isn't available from the DataTypes.

Most functions don't use the schema nullability to see if they can use a separate fast path but instead use the actual null counts  ont he arrays (e.g `Array::null_count` https://docs.rs/arrow/latest/arrow/array/trait.Array.html#method.null_count)

I wonder if that would work for your usecase?

**Comment by joseph-isaacs** [comment]
Unfortunately not.

**Comment by alamb** [comment]
🤔  It seems like the more general thing to do would be to change `return-type` to return a `Field` rather than a `DataType`  -- maybe we could add something backwards compatible

**Comment by findepi** [comment]
`Field` has a name attribute, which function top level return value doesn't have (function result is unnamed).
Top level nullability seems reported by `ScalarUDFImpl::is_nullable`.
For complex types (lists, structs), they already involve fields, and those fields can be marked nullable or non-null.

**Comment by joseph-isaacs** [comment]
It looks like [is_nullable](https://docs.rs/datafusion/latest/datafusion/logical_expr/trait.ScalarUDFImpl.html#method.is_nullable) & `return_type` can be used to specific the datatype and nullability of the scalar udf.

**Comment by joseph-isaacs** [comment]
It would be great to figure out a way to pass this info the invoke at execution time. Do you think any of the suggestion in the initial post would be a good idea?

**Comment by alamb** [comment]
> It would be great to figure out a way to pass this nullability info (or return) the invoke at execution time. Do you think any of the suggestion in the initial post would be a good idea?

I worry about making too many different combinations of invoke / invoke_no_args / invoke_nullable, etc

> 
> It seems like the nullability wasn't fully considered for function since the `is_nullable` and `return_type_from_exprs` where added later, would be great have this info throughout the trait.

I agree nullability was not fully considered and having a proper nullability story would be good. 

I think one thing that happened is that much of DataFusion's code assumes the input can be nullable and then switches to special case no-null code at runtime based on [`Array::null_count`](https://docs.rs/arrow/latest/arrow/array/trait.Array.html#method.null_count) rather than relying on the schema 

That made ensuring that the schema's nullability aligned wasn't as important and so there are a bunch of places which are fast and loose with it

**Comment by joseph-isaacs** [comment]
Any chance we can make a decision as if this will be supported or not?

**Comment by alamb** [comment]
I think there is a compelling story to be made if there is no way to implement a feature with the current API

So what I would recommend is 
1. Create a PR with an example showing what you are trying to do and any API change you want to propose

Then we can perhaps have a more specific conversation about the specific APIs and how else you might be able to solve your challenge.

**Comment by joseph-isaacs** [comment]
https://github.com/spiraldb/datafusion/pull/1/files

**Comment by alamb** [comment]
> Cool, https://github.com/spiraldb/datafusion/pull/1/files

Thanks!

Is there any way you can add an example of the `pack` UDF that shows how this API would be used?

**Comment by joseph-isaacs** [comment]
Here is a example of the usage https://github.com/apache/datafusion/pull/13289

**Comment by joseph-isaacs** [comment]
Here is the draft PR https://github.com/apache/datafusion/pull/13290

**Comment by gatesn** [comment]
I think the objection is that by all understanding, `simplification` is an optional optimization step. The result of an expression shouldn't change based on whether it was run.

In your proposal, simplification is _required_ for a correct result to be produced.

**Comment by findepi** [comment]
That's a good point, @gatesn!

What if we had similar (same?) operation, but with different guarantees (ie mandatory step). Could be called "prepare" for example. Would it help?

BTW, I would expect `simplify()` to be guaranteed. For example, if `simplify` replaces my (hypothetical) my_like_udf with eg regexp_like, i shouldn't need to provide implementation of the my_like_udf. If it's not supposed to be called, it would be dead code, useless impl burden. Does it make sense?

**Comment by gatesn** [comment]
Yes that makes sense. 

Perhaps some of the conflict comes from the fact that we're using DataFusions expression library _without_ the LogicalPlan or any of the relational algebra. It is mostly working OK, we have found a few structs that could do with being made public! Is this use-case intended to be supported by DataFusion? Or should we be wrapping everything up in a LogicalPlan for execution?

I mention this, because if we are expected to go via LogicalPlans for execution, then I would think that simplification is far more guaranteed to run than if we construct and evaluate the expressions ourselves.

As a side note, all this annoyance ultimately comes from nullability living inside a Field instead of inside a DataType. But this is an Arrow decision we have to live with 🤷 ([or do we...?](https://github.com/apache/datafusion/issues/12622))

I'm not quite sure why adding a new `prepare` function would be less disruptive to the API than passing in the return_type to invoke? Perhaps making it `invoke(args: InvokeArgs)` would protect against future unforeseen API breaks?

**Comment by alamb** [comment]
> As a side note, all this annoyance ultimately comes from nullability living inside a Field instead of inside a DataType. But this is an Arrow decision we have to live with 🤷 ([or do we...?](https://github.com/apache/datafusion/issues/12622))

Yes I agree. 

> 
> I'm not quite sure why adding a new `prepare` function would be less disruptive to the API than passing in the return_type to invoke? Perhaps making it `invoke(args: InvokeArgs)` would protect against future unforeseen API breaks?

One reason we are likely a bit worried about changing the API again is that we are in the process of changing invoke already:
- https://github.com/apache/datafusion/issues/13238

So churning it again will be annoying downstream. That being said, I think it may be worth it if we follow your suggestion to make it `args: InvokeArgs`. This helps to:
1. to protect against other API changes
2. Is consistent with [`AccumulatorArgs`](https://docs.rs/datafusion/latest/datafusion/logical_expr/function/struct.AccumulatorArgs.html)

**Comment by joseph-isaacs** [comment]
I guess changing the API again soon might reduce the number of people that migrate twice?

**Comment by alamb** [comment]
> I guess changing the API again soon might reduce the number of people that migrate twice?

Unfortunately, since we just released DataFusion 43.0.0 the new API has been released now (though the old API is marked as deprecated which I think eases the upgrade pain -- it gives them a release to do the upgrade)
